protocol/communities/roles_authorization.go

```diff
@@ -56,36 +60,51 @@ var rolesToAuthorizedPermissionTypes = map[protobuf.CommunityMember_Roles][]prot
 
 func canRolesPerformEvent(roles []protobuf.CommunityMember_Roles, eventType protobuf.CommunityEvent_EventType) bool {
 	for _, role := range roles {
-		authorizedEventTypes := rolesToAuthorizedEventTypes[role]
-		for _, authorizedEventType := range authorizedEventTypes {
-			if authorizedEventType == eventType {
-				return true
-			}
+		if slices.Contains(rolesToAuthorizedEventTypes[role], eventType) {
+			return true
 		}
 	}
 	return false
 }
 
 func canRolesModifyPermission(roles []protobuf.CommunityMember_Roles, permissionType protobuf.CommunityTokenPermission_Type) bool {
 	for _, role := range roles {
-		authorizedPermissionTypes := rolesToAuthorizedPermissionTypes[role]
-		for _, authorizedPermissionType := range authorizedPermissionTypes {
-			if authorizedPermissionType == permissionType {
-				return true
-			}
+		if slices.Contains(rolesToAuthorizedPermissionTypes[role], permissionType) {
+			return true
 		}
 	}
 	return false
 }
 
-func RolesAuthorizedToPerformEvent(roles []protobuf.CommunityMember_Roles, event *CommunityEvent) bool {
-	if !canRolesPerformEvent(roles, event.Type) {
+func canRolesKickOrBanMember(senderRoles []protobuf.CommunityMember_Roles, memberRoles []protobuf.CommunityMember_Roles) bool {
+	if slices.Contains(senderRoles, protobuf.CommunityMember_ROLE_OWNER) {
+		return true
+	}
+
+	if (slices.Contains(senderRoles, protobuf.CommunityMember_ROLE_ADMIN) ||
+		slices.Contains(senderRoles, protobuf.CommunityMember_ROLE_TOKEN_MASTER)) &&
+		!(slices.Contains(memberRoles, protobuf.CommunityMember_ROLE_ADMIN) ||
+			slices.Contains(memberRoles, protobuf.CommunityMember_ROLE_TOKEN_MASTER) ||
+			slices.Contains(memberRoles, protobuf.CommunityMember_ROLE_OWNER)) {
+		return true
+	}
```

Comment:
This is a bit more complex than this.
Admin can only kick normal members.
TokenMaster can also kick Admins
Owner can kick everyone.

Reply:
fixed